firedrake/slate/slac.py

```diff
@@ -0,0 +1,473 @@
+        """Returns the Eigen::Matrix declaration of the tensor."""
+
+        if len(shape) == 1:
+            rows = shape[0]
+            cols = 1
+        else:
+            if not len(shape) == 2:
+                raise NotImplementedError("%d-rank tensors are not currently supported." % len(shape))
+            rows = shape[0]
+            cols = shape[1]
+        if cols != 1:
+            order = ", Eigen::RowMajor"
+        else:
+            order = ""
+
+        return "Eigen::Matrix<double, %d, %d%s>" % (rows, cols, order)
+
+    def map_type(matrix):
+        """Returns an eigen map to output the resulting matrix
+        in an appropriate data array."""
+
+        return "Eigen::Map<%s >" % matrix
+
+    # Compile integrals associated with a temporary using TSFC
+    @singledispatch
+    def get_kernel_expr(expr):
+        """Retrieves the TSFC kernel function for a specific tensor."""
+
+        raise NotImplementedError("Expression of type %s not supported.",
+                                  type(expr).__name__)
+
+    @get_kernel_expr.register(Scalar)
+    @get_kernel_expr.register(Vector)
+    @get_kernel_expr.register(Matrix)
+    def get_kernel_expr_tensor(expr):
+        """Compile integral forms using TSFC form compiler."""
+
+        # No need to store identical variables more than once
+        if expr not in temps.keys():
+            sym = "T%d" % len(temps)
+            temp = ast.Symbol(sym)
+            temp_type = mat_type(expr.shape)
+            temps[expr] = temp
+            statements.append(ast.Decl(temp_type, temp))
+
+            # Compile integrals using TSFC
+            integrals = expr.form.integrals()
+            kernel_exprs[expr] = []
+            mapper = RemoveRestrictions()
+            for i, it in enumerate(integrals):
+                typ = it.integral_type()
+                form = Form([it])
+                prefix = "subkernel%d_%d_%s_" % (len(kernel_exprs), i, typ)
+
+                if typ == "interior_facet":
+                    # Reconstruct "interior_facet" integrals to be of type "exterior_facet."
+                    # This is because we are performing `local` operations, iterating over cells.
+                    # An "interior_facet" is locally an "exterior_facet."
+                    newit = map_integrand_dags(mapper, it)
+                    newit = newit.reconstruct(integral_type="exterior_facet")
+                    form = Form([newit])
+
+                tsfc_compiled_form = firedrake.tsfc_interface.compile_form(form, prefix)
+                kernel_exprs[expr].append((typ, tsfc_compiled_form))
+        return
+
+    @get_kernel_expr.register(UnaryOp)
+    @get_kernel_expr.register(BinaryOp)
+    @get_kernel_expr.register(Transpose)
+    @get_kernel_expr.register(Inverse)
+    def get_kernel_expr_ops(expr):
+        """Map operands of expressions into
+        `get_kernel_expr` recursively."""
+
+        map(get_kernel_expr, expr.operands)
+        return
```

Comment:
You're right. I think your suggestion here and below (concerning the transformer) will clean up the `compile_slate_expression` function and make it much more easy to read.